### ISSUE TITLE
v1.29.1 concurrent map read and map write

### ISSUE THREAD

**Issue by luricheng** [IBM/sarama#3018]
##### Description

github.com/!shopify/sarama@v1.29.1/balance_strategy.go
partitionMovements.Movements will caused concurrent map read and map write

##### Versions

| Sarama | Kafka | Go  |
| ------ | ----- | --- |
|    v1.29.1    |       |   1.16  |

##### Configuration

```go

```

##### Logs

<details><summary>logs: CLICK ME</summary>
<p>

```
fatal error: concurrent map read and map write

goroutine 449 [running]:
runtime.throw(0x2753b0a, 0x21)
        /usr/go/src/runtime/panic.go:1117 +0x72 fp=0xc004fd6738 sp=0xc004fd6708 pc=0x43ba52
runtime.mapaccess2(0x20d1280, 0xc02594f740, 0xc004fd67d0, 0x20, 0xc01dd589a0)
        /usr/go/src/runtime/map.go:469 +0x255 fp=0xc004fd6778 sp=0xc004fd6738 pc=0x412d15
github.com/Shopify/sarama.(*partitionMovements).getTheActualPartitionToBeMoved(0x45c9090, 0xc01dcc0f00, 0x20, 0xc000000008, 0xc027c27f80, 0x7a, 0xc027c27e80, 0x7a, 0x0, 0x19e394a, ...)
        /root/go/pkg/mod/github.com/!shopify/sarama@v1.29.1/balance_strategy.go:965 +0xc5 fp=0xc004fd6878 sp=0xc004fd6778 pc=0x19e6de5
github.com/Shopify/sarama.(*stickyBalanceStrategy).reassignPartition(0x45c9090, 0xc01dcc0f00, 0x20, 0xc000000008, 0xc02594f7a0, 0xc02e4d40c0, 0x2, 0x2, 0xc004fd7408, 0xc027c27e80, ...)
        /root/go/pkg/mod/github.com/!shopify/sarama@v1.29.1/balance_strategy.go:562 +0xe6 fp=0xc004fd6910 sp=0xc004fd6878 pc=0x19e27c6
github.com/Shopify/sarama.(*stickyBalanceStrategy).reassignPartitionToNewConsumer(0x45c9090, 0xc01dcc0f00, 0x20, 0x8, 0xc02594f7a0, 0xc02e4d40c0, 0x2, 0x2, 0xc004fd7408, 0xc004fd7438, ...)
        /root/go/pkg/mod/github.com/!shopify/sarama@v1.29.1/balance_strategy.go:552 +0x1dd fp=0xc004fd69b0 sp=0xc004fd6910 pc=0x19e267d
github.com/Shopify/sarama.(*stickyBalanceStrategy).performReassignments(0x45c9090, 0xc01dda8900, 0x14, 0x20, 0xc02594f7a0, 0xc02594fe90, 0xc02e4d40c0, 0x2, 0x2, 0xc004fd7438, ...)
        /root/go/pkg/mod/github.com/!shopify/sarama@v1.29.1/balance_strategy.go:535 +0x4f8 fp=0xc004fd6ac0 sp=0xc004fd69b0 pc=0x19e2038
github.com/Shopify/sarama.(*stickyBalanceStrategy).balance(0x45c9090, 0xc02594f7a0, 0xc02594fe90, 0xc01dda8900, 0x14, 0x20, 0x0, 0x0, 0x0, 0xc02e4d4040, ...)
        /root/go/pkg/mod/github.com/!shopify/sarama@v1.29.1/balance_strategy.go:327 +0x88c fp=0xc004fd7178 sp=0xc004fd6ac0 pc=0x19e016c
github.com/Shopify/sarama.(*stickyBalanceStrategy).Plan(0x45c9090, 0xc02594f6e0, 0xc02594f710, 0x20, 0xc01dcdd6a0, 0xa)
        /root/go/pkg/mod/github.com/!shopify/sarama@v1.29.1/balance_strategy.go:253 +0x1433 fp=0xc004fd7bc8 sp=0xc004fd7178 pc=0x19df2d3
github.com/Shopify/sarama.(*consumerGroup).balance(0xc00e85ab40, 0xc02594f6e0, 0x0, 0x0, 0x2)
        /root/go/pkg/mod/github.com/!shopify/sarama@v1.29.1/consumer_group.go:379 +0x351 fp=0xc004fd7d48 sp=0xc004fd7bc8 pc=0x1a03b71
github.com/Shopify/sarama.(*consumerGroup).newSession(0xc00e85ab40, 0x2aa8dd0, 0xc00c128880, 0xc00e705e80, 0x2, 0x2, 0x2aa0fb0, 0xc00e855380, 0x4, 0xc021194f80, ...)
        /root/go/pkg/mod/github.com/!shopify/sarama@v1.29.1/consumer_group.go:251 +0x8f6 fp=0xc004fd7e68 sp=0xc004fd7d48 pc=0x1a02c56
github.com/Shopify/sarama.(*consumerGroup).Consume(0xc00e85ab40, 0x2aa8dd0, 0xc00c128880, 0xc00e705e80, 0x2, 0x2, 0x2aa0fb0, 0xc00e855380, 0x0, 0x0)
        /root/go/pkg/mod/github.com/!shopify/sarama@v1.29.1/consumer_group.go:169 +0x18e fp=0xc004fd7f18 sp=0xc004fd7e68 pc=0x1a01d8e
git.code.oa.com/trpc-go/trpc-database/kafka.(*ServerTransport).ListenAndServe.func1(0x2aa1850, 0xc00e85ab40, 0x2aa8dd0, 0xc00c128880, 0xc0001ee9c0, 0x2aa0fb0, 0xc00e855380)
        /root/go/pkg/mod/git.code.oa.com/trpc-go/trpc-database/kafka@v0.2.9/server_transport.go:93 +0xdf fp=0xc004fd7fa8 sp=0xc004fd7f18 pc=0x1ac6f5f
runtime.goexit()
        /usr/go/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc004fd7fb0 sp=0xc004fd7fa8 pc=0x475c01
```

</p>
</details>

**Comment by puellanivis** [comment]
Does this issue persist in v1.43.3?

**Comment by luricheng** [comment]
> Does this issue persist in v1.43.3?

Sorry, I don't know. This problem is sporadic. I have used this version of the code for a long time, but I only encountered it once today and I can't find any related issues.

**Comment by dnwe** [comment]
This was fixed in 1.40

https://github.com/IBM/sarama/pull/2453
